File: scripts/rust/src/review_inherent_plus_trait_impl.py
#!/usr/bin/env python3
"""
Detect structs that have BOTH inherent impl AND trait impl blocks.

Pattern to find:
- impl<T> StructName<T> { ... }         // Inherent impl
- impl<T> TraitName<T> for StructName   // Trait impl

Most structs should have ONLY trait impl, not both.
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import re
import sys
from pathlib import Path
from collections import defaultdict

# Add lib directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'lib'))
from review_utils import ReviewContext, create_review_parser


def analyze_file(filepath, context):
    """Find structs with both inherent and trait impls."""
    try:
        with open(filepath, 'r') as f:
            lines = f.readlines()
    except Exception:
        return []
    
    # Track struct_name -> {'inherent': [line_nums], 'traits': {trait_name: [line_nums]}}
    struct_impls = defaultdict(lambda: {'inherent': [], 'traits': defaultdict(list)})
    
    for i, line in enumerate(lines, 1):
        # Match inherent impl: impl<...> StructName<...> {
        inherent_match = re.match(r'\s*impl(?:<[^>]+>)?\s+(\w+)(?:<[^>]*>)?\s*(?:where\s+|\{)', line)
        if inherent_match and ' for ' not in line:
            struct_name = inherent_match.group(1)
            struct_impls[struct_name]['inherent'].append(i)
        
        # Match trait impl: impl<...> TraitName<...> for StructName<...>
        trait_match = re.match(r'\s*impl(?:<[^>]+>)?\s+(\w+)(?:<[^>]*>)?\s+for\s+(\w+)', line)
        if trait_match:
            trait_name = trait_match.group(1)
            struct_name = trait_match.group(2)
            
            # Skip standard traits
            standard_traits = {
                'Debug', 'Clone', 'Copy', 'PartialEq', 'Eq', 'PartialOrd', 'Ord',
                'Hash', 'Display', 'Default', 'From', 'Into', 'AsRef', 'AsMut',
                'Deref', 'DerefMut', 'Drop', 'Iterator', 'IntoIterator',
                'Send', 'Sync', 'Sized', 'Unpin'
            }
            
            if trait_name not in standard_traits:
                struct_impls[struct_name]['traits'][trait_name].append(i)
    
    # Find structs with BOTH inherent AND trait impls
    violations = []
    for struct_name, impls in struct_impls.items():
        if impls['inherent'] and impls['traits']:
            violations.append({
                'struct': struct_name,
                'inherent_lines': impls['inherent'],
                'traits': dict(impls['traits'])
            })
    
    return violations


def main():
    parser = create_review_parser(
        description="Detect structs with both inherent impl and trait impl (should have trait impl only)"
    )
    args = parser.parse_args()
    context = ReviewContext(args)

    # Only check src/ files
    src_dir = context.repo_root / 'src'
    if not src_dir.exists():
        print("✗ No src/ directory found")
        return 1
    
    files = list(src_dir.rglob('*.rs'))
    
    all_violations = {}
    
    for filepath in sorted(files):
        violations = analyze_file(filepath, context)
        if violations:
            all_violations[filepath] = violations
    
    if not all_violations:
        print("\n✓ All structs use trait impl only (no inherent+trait duplication)!")
        return 0
    
    # Count total violations
    total_count = sum(len(v) for v in all_violations.values())
    
    print(f"✗ Inherent + Trait Impl Pattern: {total_count} struct(s)\n")
    print("Structs should use TRAIT impl only, not both inherent impl and trait impl.\n")
    print("="*80)
    
    for filepath, file_violations in sorted(all_violations.items()):
        rel_path = context.relative_path(filepath)
        print(f"\n{rel_path}:")
        
        for violation in file_violations:
            struct_name = violation['struct']
            inherent_lines = violation['inherent_lines']
            traits = violation['traits']
            
            print(f"  {struct_name}:")
            print(f"    Inherent impl at line(s): {', '.join(map(str, inherent_lines))}")
            for trait_name, trait_lines in sorted(traits.items()):
                print(f"    Trait impl {trait_name} at line(s): {', '.join(map(str, trait_lines))}")
    
    print(f"\n{'='*80}")
    print(f"Total: {total_count} struct(s) with both inherent and trait impls")
    print("\nRecommendation: Remove inherent impl, keep only trait impl.")
    
    return 1 if all_violations else 0


if __name__ == '__main__':
    sys.exit(main())

File: src/bin/review_inherent_plus_trait_impl.rs
//! Review: Inherent + Trait Impl Pattern
//!
//! Detects structs that have BOTH inherent impl AND trait impl blocks.
//! Most structs should have ONLY trait impl, not both.
//!
//! Binary: rusticate-review-inherent-plus-trait-impl

use anyhow::Result;
use ra_ap_syntax::{ast::{self, AstNode}, SyntaxKind, SourceFile, Edition};
use rusticate::{StandardArgs, find_rust_files};
use std::collections::HashMap;
use std::path::Path;
use std::time::Instant;

const STANDARD_TRAITS: &[&str] = &[
    "Debug", "Clone", "Copy", "PartialEq", "Eq", "PartialOrd", "Ord",
    "Hash", "Display", "Default", "From", "Into", "AsRef", "AsMut",
    "Deref", "DerefMut", "Drop", "Iterator", "IntoIterator",
    "Send", "Sync", "Sized", "Unpin"
];


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_inherent_plus_trait_impl.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
#[derive(Debug)]
struct Violation {
    struct_name: String,
    inherent_lines: Vec<usize>,
    traits: HashMap<String, Vec<usize>>,
}

fn analyze_file(file_path: &Path) -> Vec<Violation> {
    let content = match std::fs::read_to_string(file_path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let parsed = SourceFile::parse(&content, Edition::Edition2021);
    let tree = parsed.tree();
    let root = tree.syntax();

    // Track struct_name -> {'inherent': [lines], 'traits': {trait_name: [lines]}}
    let mut struct_impls: HashMap<String, (Vec<usize>, HashMap<String, Vec<usize>>)> = HashMap::new();

    for node in root.descendants() {
        if node.kind() == SyntaxKind::IMPL {
            if let Some(impl_node) = ast::Impl::cast(node.clone()) {
                let line_num = content[..node.text_range().start().into()]
                    .chars()
                    .filter(|&c| c == '\n')
                    .count() + 1;

                if let Some(self_ty) = impl_node.self_ty() {
                    let self_ty_str = self_ty.to_string();
                    // Extract base type name
                    let struct_name = if let Some(pos) = self_ty_str.find('<') {
                        self_ty_str[..pos].trim().to_string()
                    } else {
                        self_ty_str.trim().to_string()
                    };

                    if impl_node.trait_().is_none() {
                        // Inherent impl
                        struct_impls.entry(struct_name).or_insert((Vec::new(), HashMap::new())).0.push(line_num);
                    } else {
                        // Trait impl
                        if let Some(trait_ref) = impl_node.trait_() {
                            let trait_str = trait_ref.to_string();
                            // Extract trait name (last component)
                            let trait_name = trait_str.split("::").last().unwrap_or(&trait_str);
                            
                            // Remove generics
                            let trait_name = if let Some(pos) = trait_name.find('<') {
                                trait_name[..pos].trim().to_string()
                            } else {
                                trait_name.trim().to_string()
                            };

                            // Skip standard traits
                            if !STANDARD_TRAITS.contains(&trait_name.as_str()) {
                                struct_impls.entry(struct_name).or_insert((Vec::new(), HashMap::new()))
                                    .1.entry(trait_name).or_default().push(line_num);
                            }
                        }
                    }
                }
            }
        }
    }

    // Find structs with BOTH inherent AND trait impls
    let mut violations = Vec::new();
    for (struct_name, (inherent_lines, traits)) in struct_impls {
        if !inherent_lines.is_empty() && !traits.is_empty() {
            violations.push(Violation {
                struct_name,
                inherent_lines,
                traits,
            });
        }
    }

    violations
}

fn main() -> Result<()> {
    let start_time = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();

    // Only check src/ files
    let src_files: Vec<_> = find_rust_files(&args.paths)
        .into_iter()
        .filter(|p| p.starts_with(base_dir.join("src")))
        .collect();

    let mut all_violations: HashMap<_, Vec<Violation>> = HashMap::new();

    for file_path in &src_files {
        let violations = analyze_file(file_path);
        if !violations.is_empty() {
            all_violations.insert(file_path.clone(), violations);
        }
    }

    if all_violations.is_empty() {
        log!("\n✓ All structs use trait impl only (no inherent+trait duplication)!");
        let elapsed = start_time.elapsed();
        eprintln!("\nCompleted in {}ms", elapsed.as_millis());
        return Ok(());
    }

    let total_count: usize = all_violations.values().map(|v| v.len()).sum();

    log!("✗ Inherent + Trait Impl Pattern: {} struct(s)\n", total_count);
    log!("Structs should use TRAIT impl only, not both inherent impl and trait impl.\n");
    log!("{}", "=".repeat(80));

    for (file_path, file_violations) in all_violations.iter() {
        let rel_path = file_path.strip_prefix(&base_dir).unwrap_or(file_path);
        log!("\n{}:", rel_path.display());

        for violation in file_violations {
            log!("  {}:", violation.struct_name);
            let inherent_str = violation.inherent_lines.iter()
                .map(|l| l.to_string())
                .collect::<Vec<_>>()
                .join(", ");
            log!("    Inherent impl at line(s): {}", inherent_str);
            
            let mut trait_names: Vec<_> = violation.traits.keys().collect();
            trait_names.sort();
            for trait_name in trait_names {
                let lines = &violation.traits[trait_name];
                let lines_str = lines.iter().map(|l| l.to_string()).collect::<Vec<_>>().join(", ");
                log!("    Trait impl {} at line(s): {}", trait_name, lines_str);
            }
        }
    }

    log!("\n{}", "=".repeat(80));
    log!("Total: {} struct(s) with both inherent and trait impls", total_count);
    log!("\nRecommendation: Remove inherent impl, keep only trait impl.");

    let elapsed = start_time.elapsed();
    eprintln!("\nCompleted in {}ms", elapsed.as_millis());

    Ok(())
}

